python/dolfinx/fem/linearvariationalsolver.py

```diff
@@ -0,0 +1,80 @@
+# -*- coding: utf-8 -*-
+# Copyright (C) 2020 Jørgen S. Dokken
+#
+# This file is part of DOLFINX (https://www.fenicsproject.org)
+#
+# SPDX-License-Identifier:    LGPL-3.0-or-later
+import typing
+
+from petsc4py import PETSc
+from ufl.equation import Equation
+from dolfinx import fem
+
+
+class LinearVariationalSolver():
+    assemble_b = True
+
+    def __init__(self, variational_problem: Equation, u: fem.Function,
+                 bcs: typing.List[fem.DirichletBC] = [], form_compiler_parameters={}):
+        """
+        Initialize the linear variational solver. By initialization we mean creating the underlying PETSc solver,
+        and the linear algebra structures (the matrix and vector), that can be reused for multiple solve calls.
+
+        .. code-block:: python
+
+            solver = LinearVariationalSolver(a==L, u, [bc0, bc1], form_compiler_parameters={"optimize": True})
+
+        """
+        if (len(variational_problem.lhs.arguments()) != 2):
+            raise ValueError(
+                "Variational problem has to consist of a bi-linear form,"
+                + " defined with a `ufl.TrialFunction` and `ufl.TestFunction`.")
+        # If a == 0 is inserted just assume that rhs is 0.
+        if (variational_problem.rhs == 0):
+            self.assemble_b = False
+            self.b = u.vector.copy()
```

Comment:
The RHS vector should be created by the matrix object to ensure a consistent layout.